File: src/main/java/pw/switchcraft/plethora/gameplay/registry/PlethoraMethodRegistration.java
package pw.switchcraft.plethora.gameplay.registry;

import pw.switchcraft.plethora.api.method.IMethod;
import pw.switchcraft.plethora.api.method.IMethodCollection;
import pw.switchcraft.plethora.api.method.IMethodRegistry;
import pw.switchcraft.plethora.api.module.IModuleContainer;
import pw.switchcraft.plethora.gameplay.modules.glasses.GlassesMethods;
import pw.switchcraft.plethora.gameplay.modules.glasses.canvas.Canvas2dMethods;
import pw.switchcraft.plethora.gameplay.modules.glasses.canvas.Canvas3dMethods;
import pw.switchcraft.plethora.gameplay.modules.glasses.objects.*;
import pw.switchcraft.plethora.gameplay.modules.glasses.objects.object2d.MultiPoint2d;
import pw.switchcraft.plethora.gameplay.modules.glasses.objects.object2d.MultiPointResizable2d;
import pw.switchcraft.plethora.gameplay.modules.glasses.objects.object2d.Positionable2d;
import pw.switchcraft.plethora.gameplay.modules.glasses.objects.object2d.Rectangle;
import pw.switchcraft.plethora.gameplay.modules.glasses.objects.object3d.*;
import pw.switchcraft.plethora.gameplay.modules.introspection.IntrospectionMethods;
import pw.switchcraft.plethora.gameplay.modules.kinetic.KineticMethods;
import pw.switchcraft.plethora.gameplay.modules.laser.LaserMethods;
import pw.switchcraft.plethora.gameplay.modules.scanner.ScannerMethods;
import pw.switchcraft.plethora.gameplay.modules.sensor.SensorMethods;
import pw.switchcraft.plethora.integration.CoreMethods;
import pw.switchcraft.plethora.integration.GetMetadataMethod;

import static pw.switchcraft.plethora.Plethora.MOD_ID;

final class PlethoraMethodRegistration {
    static void registerMethods(IMethodRegistry r) {
        // Core methods
        method(r, Object.class, new GetMetadataMethod());
        method(r, IModuleContainer.class, CoreMethods.LIST_MODULES);
        method(r, IModuleContainer.class, CoreMethods.HAS_MODULE);
        method(r, IModuleContainer.class, CoreMethods.FILTER_MODULES);
        method(r, IMethodCollection.class, CoreMethods.GET_DOCS);

        // Modules
        moduleMethod(r, "introspection:getID", IntrospectionMethods.GET_ID);
        moduleMethod(r, "introspection:getName", IntrospectionMethods.GET_NAME);
        moduleMethod(r, "introspection:getMetaOwner", IntrospectionMethods.GET_META_OWNER);

        moduleMethod(r, "kinetic:launch", KineticMethods.LAUNCH);

        moduleMethod(r, "laser:fire", LaserMethods.FIRE);

        moduleMethod(r, "sensor:sense", SensorMethods.SENSE);
        moduleMethod(r, "sensor:getMetaByID", SensorMethods.GET_META_BY_ID);
        moduleMethod(r, "sensor:getMetaByName", SensorMethods.GET_META_BY_NAME);

        moduleMethod(r, "scanner:sense", ScannerMethods.SCAN);
        moduleMethod(r, "scanner:getBlockMeta", ScannerMethods.GET_BLOCK_META);

        // Overlay glasses
        moduleMethod(r, "glasses:canvas", GlassesMethods.GET_CANVAS);
        moduleMethod(r, "glasses:canvas3d", GlassesMethods.GET_CANVAS_3D);

        method(r, ObjectGroup.class, GlassesMethods.CLEAR);
        method(r, BaseObject.class, GlassesMethods.REMOVE);
        methods(r, Positionable2d.class, Positionable2d.GET_POSITION, Positionable2d.SET_POSITION);
        methods(r, Colourable.class, Colourable.GET_COLOUR, Colourable.GET_COLOR, Colourable.SET_COLOUR,
            Colourable.SET_COLOR, Colourable.GET_ALPHA, Colourable.SET_ALPHA);
        methods(r, Rectangle.class, Rectangle.GET_SIZE, Rectangle.SET_SIZE);
        methods(r, MultiPoint2d.class, MultiPoint2d.GET_POINT, MultiPoint2d.SET_POINT);
        methods(r, MultiPointResizable2d.class, MultiPointResizable2d.GET_POINT_COUNT,
            MultiPointResizable2d.REMOVE_POINT, MultiPointResizable2d.INSERT_POINT);
        methods(r, Scalable.class, Scalable.GET_SCALE, Scalable.SET_SCALE);
        methods(r, TextObject.class, TextObject.GET_TEXT, TextObject.SET_TEXT, TextObject.SET_SHADOW,
            TextObject.HAS_SHADOW, TextObject.GET_LINE_HEIGHT, TextObject.SET_LINE_HEIGHT);
        methods(r, ItemObject.class, ItemObject.GET_ITEM, ItemObject.SET_ITEM);

        methods(r, ObjectGroup.Group2d.class, Canvas2dMethods.ADD_RECTANGLE, Canvas2dMethods.ADD_LINE,
            Canvas2dMethods.ADD_DOT, Canvas2dMethods.ADD_TEXT, Canvas2dMethods.ADD_TRIANGLE,
            Canvas2dMethods.ADD_POLYGON, Canvas2dMethods.ADD_LINES, Canvas2dMethods.ADD_ITEM,
            Canvas2dMethods.ADD_GROUP);
        methods(r, ObjectGroup.Frame2d.class, Canvas2dMethods.GET_SIZE);

        methods(r, Positionable3d.class, Positionable3d.GET_POSITION, Positionable3d.SET_POSITION);
        methods(r, Rotatable3d.class, Rotatable3d.GET_ROTATION, Rotatable3d.SET_ROTATION);
        methods(r, DepthTestable.class, DepthTestable.IS_DEPTH_TESTED, DepthTestable.SET_DEPTH_TESTED);
        methods(r, Box.class, Box.GET_SIZE, Box.SET_SIZE);
        methods(r, ObjectRoot3d.class, ObjectRoot3d.RECENTER);
        methods(r, ObjectGroup.Origin3d.class, Canvas3dMethods.CREATE);
        methods(r, ObjectGroup.Group3d.class, Canvas3dMethods.ADD_BOX);
    }

    private static <T> void method(IMethodRegistry r, String name, Class<T> target, IMethod<T> method) {
        r.registerMethod(MOD_ID, name, target, method);
    }

    private static <T> void method(IMethodRegistry r, Class<T> target, IMethod<T> method) {
        r.registerMethod(MOD_ID, method.getName(), target, method);
    }

    @SafeVarargs
    private static <T> void methods(IMethodRegistry r, Class<T> target, IMethod<T>... methods) {
        for (IMethod<T> method : methods) {
            method(r, target, method);
        }
    }

    private static void moduleMethod(IMethodRegistry r, String name, IMethod<IModuleContainer> method) {
        method(r, name, IModuleContainer.class, method);
    }
}

File: src/main/java/io/sc3/plethora/gameplay/registry/PlethoraMethodRegistration.kt
package io.sc3.plethora.gameplay.registry

import io.sc3.plethora.Plethora.modId
import io.sc3.plethora.api.method.IMethod
import io.sc3.plethora.api.method.IMethodCollection
import io.sc3.plethora.api.method.IMethodRegistry
import io.sc3.plethora.api.module.IModuleContainer
import io.sc3.plethora.gameplay.modules.glasses.GlassesMethods
import io.sc3.plethora.gameplay.modules.glasses.canvas.Canvas2dMethods
import io.sc3.plethora.gameplay.modules.glasses.canvas.Canvas3dMethods
import io.sc3.plethora.gameplay.modules.glasses.objects.*
import io.sc3.plethora.gameplay.modules.glasses.objects.ObjectGroup.*
import io.sc3.plethora.gameplay.modules.glasses.objects.object2d.MultiPoint2d
import io.sc3.plethora.gameplay.modules.glasses.objects.object2d.MultiPointResizable2d
import io.sc3.plethora.gameplay.modules.glasses.objects.object2d.Positionable2d
import io.sc3.plethora.gameplay.modules.glasses.objects.object2d.Rectangle2d
import io.sc3.plethora.gameplay.modules.glasses.objects.object3d.*
import io.sc3.plethora.gameplay.modules.introspection.IntrospectionMethods
import io.sc3.plethora.gameplay.modules.kinetic.KineticMethods
import io.sc3.plethora.gameplay.modules.laser.LaserMethods
import io.sc3.plethora.gameplay.modules.scanner.ScannerMethods
import io.sc3.plethora.gameplay.modules.sensor.SensorMethods
import io.sc3.plethora.integration.CoreMethods
import io.sc3.plethora.integration.GetMetadataMethod

internal object PlethoraMethodRegistration {
  @JvmStatic
  fun registerMethods(r: IMethodRegistry) {
    with (r) {
      // Core methods
      methods(Any::class.java, GetMetadataMethod())
      methods(IModuleContainer::class.java, CoreMethods.LIST_MODULES, CoreMethods.HAS_MODULE,
        CoreMethods.FILTER_MODULES)
      methods(IMethodCollection::class.java, CoreMethods.GET_DOCS)

      // Modules
      moduleMethod("introspection:getID", IntrospectionMethods.GET_ID)
      moduleMethod("introspection:getName", IntrospectionMethods.GET_NAME)
      moduleMethod("introspection:getMetaOwner", IntrospectionMethods.GET_META_OWNER)
      moduleMethod("kinetic:launch", KineticMethods.LAUNCH)
      moduleMethod("laser:fire", LaserMethods.FIRE)
      moduleMethod("sensor:sense", SensorMethods.SENSE)
      moduleMethod("sensor:getMetaByID", SensorMethods.GET_META_BY_ID)
      moduleMethod("sensor:getMetaByName", SensorMethods.GET_META_BY_NAME)
      moduleMethod("scanner:sense", ScannerMethods.SCAN)
      moduleMethod("scanner:getBlockMeta", ScannerMethods.GET_BLOCK_META)

      // Overlay glasses
      moduleMethod("glasses:canvas", GlassesMethods.GET_CANVAS)
      moduleMethod("glasses:canvas3d", GlassesMethods.GET_CANVAS_3D)

      methods(ObjectGroup::class.java, GlassesMethods.CLEAR)
      methods(BaseObject::class.java, GlassesMethods.REMOVE)
      methods(Positionable2d::class.java, Positionable2d.GET_POSITION, Positionable2d.SET_POSITION)
      methods(Colourable::class.java, Colourable.GET_COLOUR, Colourable.GET_COLOR, Colourable.SET_COLOUR,
        Colourable.SET_COLOR, Colourable.GET_ALPHA, Colourable.SET_ALPHA)
      methods(Rectangle2d::class.java, Rectangle2d.GET_SIZE, Rectangle2d.SET_SIZE)
      methods(MultiPoint2d::class.java, MultiPoint2d.GET_POINT, MultiPoint2d.SET_POINT)
      methods(MultiPointResizable2d::class.java, MultiPointResizable2d.GET_POINT_COUNT,
        MultiPointResizable2d.REMOVE_POINT, MultiPointResizable2d.INSERT_POINT)
      methods(Scalable::class.java, Scalable.GET_SCALE, Scalable.SET_SCALE)
      methods(TextObject::class.java, TextObject.GET_TEXT, TextObject.SET_TEXT, TextObject.SET_SHADOW,
        TextObject.HAS_SHADOW, TextObject.GET_LINE_HEIGHT, TextObject.SET_LINE_HEIGHT)
      methods(ItemObject::class.java, ItemObject.GET_ITEM, ItemObject.SET_ITEM)
      methods(Group2d::class.java, Canvas2dMethods.ADD_RECTANGLE, Canvas2dMethods.ADD_LINE, Canvas2dMethods.ADD_DOT,
        Canvas2dMethods.ADD_TEXT, Canvas2dMethods.ADD_TRIANGLE, Canvas2dMethods.ADD_POLYGON, Canvas2dMethods.ADD_LINES,
        Canvas2dMethods.ADD_ITEM, Canvas2dMethods.ADD_GROUP)
      methods(Frame2d::class.java, Canvas2dMethods.GET_SIZE)

      methods(Positionable3d::class.java, Positionable3d.GET_POSITION, Positionable3d.SET_POSITION)
      methods(Rotatable3d::class.java, Rotatable3d.GET_ROTATION, Rotatable3d.SET_ROTATION)
      methods(DepthTestable::class.java, DepthTestable.IS_DEPTH_TESTED, DepthTestable.SET_DEPTH_TESTED)
      methods(Box3d::class.java, Box3d.GET_SIZE, Box3d.SET_SIZE)
      methods(ObjectRoot3d::class.java, ObjectRoot3d.RECENTER)
      methods(Origin3d::class.java, Canvas3dMethods.CREATE)
      methods(Group3d::class.java, Canvas3dMethods.ADD_FRAME, Canvas3dMethods.ADD_BOX, Canvas3dMethods.ADD_ITEM)
    }
  }

  private fun <T> IMethodRegistry.method(name: String, target: Class<T>, method: IMethod<T>) {
    registerMethod(modId, name, target, method)
  }

  @SafeVarargs
  private fun <T> IMethodRegistry.methods(target: Class<T>, vararg methods: IMethod<T>) {
    for (method in methods) {
      registerMethod(modId, method.name, target, method)
    }
  }

  private fun IMethodRegistry.moduleMethod(name: String, method: IMethod<IModuleContainer>) {
    method(name, IModuleContainer::class.java, method)
  }
}
